ipfscluster_test.go

```diff
@@ -1530,90 +1530,91 @@ func TestClustersReplicationMinMaxNoRealloc(t *testing.T) {
 	}
 }
 
-// This test checks that repinning something that has becomed
-// underpinned does re-allocations when it's not sufficiently
-// pinned anymore
-func TestClustersReplicationMinMaxRealloc(t *testing.T) {
-	ctx := context.Background()
-	if nClusters < 5 {
-		t.Skip("Need at least 5 peers")
-	}
-
-	clusters, mock := createClusters(t)
-	defer shutdownClusters(t, clusters, mock)
-	for _, c := range clusters {
-		c.config.ReplicationFactorMin = 3
-		c.config.ReplicationFactorMax = 4
-	}
-
-	ttlDelay() // make sure metrics are in
-
-	h := test.Cid1
-	_, err := clusters[0].Pin(ctx, h, api.PinOptions{})
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	pinDelay()
-
-	p, err := clusters[0].PinGet(ctx, h)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	firstAllocations := p.Allocations
-
-	peerIDMap := make(map[peer.ID]*Cluster)
-	for _, a := range clusters {
-		peerIDMap[a.id] = a
-	}
-
-	// kill two of the allocations
-	alloc1 := peerIDMap[firstAllocations[0]]
-	alloc2 := peerIDMap[firstAllocations[1]]
-	safePeer := peerIDMap[firstAllocations[2]]
-
-	alloc1.Shutdown(ctx)
-	alloc2.Shutdown(ctx)
-
-	waitForLeaderAndMetrics(t, clusters)
-
-	// Repin - (although this might have been taken of if there was an alert
-	_, err = safePeer.Pin(ctx, h, api.PinOptions{})
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	pinDelay()
-
-	p, err = safePeer.PinGet(ctx, h)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	secondAllocations := p.Allocations
-
-	strings1 := api.PeersToStrings(firstAllocations)
-	strings2 := api.PeersToStrings(secondAllocations)
-	sort.Strings(strings1)
-	sort.Strings(strings2)
-	t.Logf("Allocs1: %s", strings1)
-	t.Logf("Allocs2: %s", strings2)
-
-	if fmt.Sprintf("%s", strings1) == fmt.Sprintf("%s", strings2) {
-		t.Error("allocations should have changed")
-	}
-
-	lenSA := len(secondAllocations)
-	expected := minInt(nClusters-2, 4)
-	if lenSA != expected {
-		t.Errorf("Insufficient reallocation, could have allocated to %d peers but instead only allocated to %d peers", expected, lenSA)
-	}
-
-	if lenSA < 3 {
-		t.Error("allocations should be more than rplMin")
-	}
-}
+// Repinning does not re-allocate anymore
```

Comment:
https://github.com/ipfs/ipfs-cluster/blob/04a5881b073dff2fcaf6ae40a63a7ff582fda2c7/cluster.go#L1385-L1389

Shouldn't it reallocate on repin?

Reply:
The idea is that if you send the same pin with the same options, and the pin is already there, the operation should just try to re-pin whenever the pin was originally (similar to recover), even if the pin is in error somewhere. This is more natural than changing the allocations (Matt from Pinana convinced me). To change the allocations the replication factor can be changed, or the item unpinned and re-added.

Logic is a little brittle though and allocate() tries to respect existing allocations, so maybe it should indeed re-allocate (and should get the same allocataions whenever the allocated peers are still there).